src/TraceEvent/TraceLog.cs

```diff
@@ -153,10 +153,199 @@ public static TraceLog OpenOrConvert(string etlOrEtlxFilePath, TraceLogOptions o
+        /// };
+        /// var session = client.StartEventPipeSession(providers, requestRundown: false);
+        /// var eventSource = TraceLog.CreateFromEventPipeSession(session, rundownSession);
+        /// eventSource.Process();
+        /// </example>
+        public static TraceLogEventSource CreateFromEventPipeSession(EventPipeSession session, EventPipeSession rundownSession = null)
+        {
+            var traceLog = new TraceLog(new EventPipeEventSource(session.EventStream));
+            traceLog.rawEventSourceToConvert.AllEvents += traceLog.OnAllEventPipeEventsRealTime;
+
+            if (rundownSession != null)
+            {
+                traceLog.ProcessInitialRundown(rundownSession);
+            }
+
             return traceLog.realTimeSource;
         }
 
+        private void ProcessInitialRundown(EventPipeSession session)
+        {
+            using (var source = new EventPipeEventSource(session.EventStream))
+            {
+                SetupInitialRundownCallbacks(source);
+
+                // Only stopping the session will cause the rundown events to be sent.
+                // However, we cannot stop it before starting to process the source.
+                // Therefore, we need attach to the first event and only then stop the session.
+                var completionSource = new TaskCompletionSource<bool>();
+                source.AllEvents += delegate (TraceEvent _)
+                {
+                    completionSource.TrySetResult(true);
+                };
+                var task = Task.Run(source.Process);
+                completionSource.Task.Wait();
+                session.Stop();
+                task.Wait();
+            }
+        }
+
+        /// <summary>
+        /// Attaches callbacks to initialize modules and methods based on the rundown events.
+        /// Unlike SetupCallbacks(), these callbacks are attached before any events are processed so
+        /// they do not rely on ForAllUnresolvedCodeAddressesInRange() and instead record all the methods.
+        /// </summary>
+        private void SetupInitialRundownCallbacks(EventPipeEventSource rawEvents)
+        {
+            rawEvents.HeadersDeserialized += delegate () { CopyHeadersFrom(rawEvents); };
+
+            var clrRundownParser = new ClrRundownTraceEventParser(rawEvents);
+            Action<ModuleLoadUnloadTraceData> onLoaderRundown = delegate (ModuleLoadUnloadTraceData data)
+            {
+                TraceProcess process = processes.GetOrCreateProcess(data.ProcessID, data.TimeStampQPC);
+                process.LoadedModules.ManagedModuleLoadOrUnload(data, false, true);
+            };
+
+            clrRundownParser.LoaderModuleDCStop += onLoaderRundown;
+            clrRundownParser.LoaderModuleDCStart += onLoaderRundown;
+
+            // TODO We don't seem to be getting any of these events, only MethodDCStopVerbose.
```

Comment:
This makes sense because rundown is limited to what's called "stop rundown" - rundown that is triggered when we stop tracing.  You can safely remove this commented out code.

Reply:
thx, removed